torchinductor/lowering.py

```diff
@@ -1313,32 +1313,62 @@ def fn(idx):
     )
 
 
+@register_lowering(aten.scatter_, type_promote=False)
```

Comment:
should this (and `scatter_add_`) be a decomposition instead?

Reply:
Fixed `scatter_add_` in the new commit. Not able to decompose `scatter_` into `scatter_reduce_`, because `reduce` can't be None for `scatter_reduce_`.

Reply:
But you can put the exact code you have below in decomposition to make sure `reduce` is not None?

Reply:
I am not sure I understand. `reduce` can be None for `scatter_`.

Reply:
Yes, but you put this function as is in decomposition.py and decorate it with @register_decomposition instead of @register_lowering, would this not work?

Reply:
No, it doesn't work. It will trigger a checking failure from aot_grad/.

Reply:
I'm assuming because AOT autograd can't handle mutation.

Reply:
I didn't really trace it but I think what happens here is because my implementation of `scatter_reduce_` is not exactly the same as the semantics of `scatter_reduce_` in PyTorch. The "official" `scatter_reduce_` does not take `None` as an input to the `reduce` parameter. So we can't decompose `scatter_` without `add` into `scatter_reduce_` when we leave the torchinductor land. I could be wrong though.

Reply:
Might be helpful to paste the error msg here,
```
...
../pytorch-release/torch/_ops.py:148: in __call__
    return self._op(*args, **kwargs or {})
../pytorch-release/torch/utils/_python_dispatch.py:74: in wrapped
    return f(self, *args, **kwargs)
../pytorch-release/torch/fx/experimental/proxy_tensor.py:290: in __torch_dispatch__
    return proxy_call(func_overload, args, kwargs)
../pytorch-release/torch/fx/experimental/proxy_tensor.py:78: in proxy_call
    return CURRENT_DECOMPOSITION_TABLE[func_overload](*args, **kwargs)
torchinductor/decomposition.py:376: in scatter
    return self.clone().scatter_(dim, index, src, **kwargs)
../pytorch-release/torch/fx/experimental/proxy_tensor.py:206: in __torch_dispatch__
    return proxy_call(func_overload, args, kwargs)
../pytorch-release/torch/fx/experimental/proxy_tensor.py:78: in proxy_call
    return CURRENT_DECOMPOSITION_TABLE[func_overload](*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ProxyTensor(tensor([[0., 0., 0.],
        [0., 0., 0.]], device='cuda:0'), proxy=Proxy(clone_default)), dim = 0, index = ProxyTensor(tensor([[0]], device='cuda:0'), proxy=Proxy(primals_2))
src = ProxyTensor(tensor([[1., 1., 1.],
        [1., 1., 1.]], device='cuda:0'), proxy=Proxy(primals_3))

    @register_decomposition([aten.scatter_])
    def scatter_(self, dim: int, index, src, *, reduce: str = None):
        if reduce == "add":
            reduce = "sum"
        elif reduce == "multiply":
            reduce = "prod"
        else:
            assert reduce is None
>       return self.scatter_reduce_(dim, index, src, reduce)
E       TypeError: scatter_reduce_(): argument 'reduce' (position 4) must be str, not NoneType
```